File: Python/product_except_self.py
class Solution:
    def productExceptSelf(self, nums):
        n = len(nums)
        result = [1] * n
        p = 1
        for i in range(n-2, -1, -1):
            print(i)
            p *= nums[i+1]
            result[i] = p

        p = 1
        for i in range(1, n):
            p *= nums[i-1]
            result[i] *= p

        return result
print(Solution().productExceptSelf([1, 2, 0, 0, 4]))

File: C++/product_except_self.cpp
class Solution {
public:
    // Time complexity: O(n)
    // Space complexity: O(1)
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans(n,1);
        ans[0] = 1;
        int p = 1;
        for (int i = n-2; i >= 0; --i) {
            p *= nums[i+1];
            ans[i] = p;
        }
        p = 1;
        for (int i = 0; i < n; ++i) {
            ans[i] *= p;
            p *= nums[i];
        }
        return ans;
    }
};
